### PR TITLE
Include values for generics and stereotypes in the output

### LINKED ISSUE #42: Parse and retain information located in Decorators (Generics and Stereotypes)
At the moment, the parser parses Generics and Stereotypes but discards their value.

Could they be included in the output ?

I'm open to making a PR if you are ok with this idea.

### PR DESCRIPTION
**Pull request by paduc**
Fixes #42 

In order to use the values inside the generics and stereotype syntax operators, I did the following:

- Added `generics_` and `stereotype_` properties to `Class`, `Interface` and `Enum` types
- Separated `Decorators` into `Stereotype` and `Generics` parsers
- `Stereotype` and `Generics` parsers return their values
- Adapted the `Class`, `Enum` and `Interface` parsers to make use of the `Stereotype` and `Generics` parsers and forward their values to the corresponding types.

Please note that the allowed declaration format for classes, interfaces and enums changed a bit (without breaking any of the examples).
Previously, the parser would allow ex `class MyClass<generic> <<stereotype>> extends ... {` (the generic and stereotypes follow each other).
After this PR, it allows ex `class MyClass<generic> extends ... <<stereotype>> {` (the generic is right after the class name and the stereotype is at the very end). This seemed better for me, as a stereotype is a sort of annotation. That is up for debate.

### CODE REVIEW

File: test/fixtures/class-types/in.plantuml [hunk 1]
```diff
@@ -9,7 +9,13 @@ abstract class AbstractList {
 
 }
 
-class Test << general >> {
+class Test < general > << stereo >> {
```
Comment (Enteee) [inline test/fixtures/class-types/in.plantuml:12]:
```suggestion
class TestGenericsAndStereotype < general > << stereo >> {
```

File: test/fixtures/class-types/in.plantuml [hunk 2]
```diff
@@ -9,7 +9,13 @@ abstract class AbstractList {
...
+}
+
+class Test < general only > {
```
Comment (Enteee) [inline test/fixtures/class-types/in.plantuml:15]:
```suggestion
class TestOnlyGenerics < general only > {
```

File: test/fixtures/class-types/in.plantuml [hunk 3]
```diff
@@ -9,7 +9,13 @@ abstract class AbstractList {
...
+}
+
+class Test<<stereo only>>{
```
Comment (Enteee) [inline test/fixtures/class-types/in.plantuml:18]:
```suggestion
class TestOnlyStereotype <<stereo only>>{
```

File: src/plantuml.pegjs [hunk 4]
```diff
@@ -205,26 +205,30 @@ NoteOf
...
     return new types.Class(
       name.name,
       name.title,
       !!isAbstract,
       removeUndefined(members),
       extends_,
-      implements_
+      implements_,
+      generics_,
+      stereotype_
```
[the remark this answers is not in the document]
Comment (paduc) [reply src/plantuml.pegjs]: Oh I didn't catch that ! Makes sense !

File: src/plantuml.pegjs [hunk 5]
```diff
@@ -489,6 +501,19 @@ RelationshipLabel
 RelationshipHidden
   = "[hidden]"
 
+
+Generics
+  = !"<<" "<" _ generics:(( !">" . )+) _ ">"
+  {
+    return generics.map(removeUndefined).join('').trim();
```
[the remark this answers is not in the document]
Comment (paduc) [reply src/plantuml.pegjs]: Awesome, didn't see this function existed !

**Comment by paduc** [comment]
Hey @Enteee thanks for you review ! I pushed two commits to fix your suggestions.

File: src/plantuml.pegjs [hunk 6]
```diff
@@ -489,6 +501,19 @@ RelationshipLabel
 RelationshipHidden
   = "[hidden]"
 
+
+Generics
+  = !"<<" "<" _ generics:(( !">" . )+) _ ">"
```
[the remark this answers is not in the document]
Comment (paduc) [reply src/plantuml.pegjs:506]: There was an overlap between genetics and stereotype. The first < of a stereotype declaration would open the generics case.

**Comment by Enteee** [comment]
@paduc I reviewd it and pushed a new commit which fixes some minor formatting issues. By the way you can run `npm run format-fix` in order to fix them automatically. There is still one open question from the review. I do now understand why you chose the design as-is, but I think it can maybe be improved. My main critique point is that the `Generics` non terminal has to contain context information (`!"<<") in order to work. Maybe that's needed, and your design is perfectly valid. But let me play around with it a bit, so I might come up with a different version.

**Comment by paduc** [comment]
@Enteee sorry about the formatting. I did run the command but there might be some conflict with my IDE lint config...

For the generics negative prefix, I don’t see how we can do without it but I’m no pegjs expert. I might learn something :)

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        "stereotype": ""
+      },
+      {
+        "name": "TestGenericsWithAngleBrackets",
+        "title": "TestGenericsWithAngleBrackets",
+        "isAbstract": false,
+        "members": [],
+        "extends_": [],
+        "implements_": [],
+        "generics": "<<G1",
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:31]: should be `"<<G1>>"`

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+      },
+      {
+        "name": "TestMultipleStereotypes",
+        "title": "TestMultipleStereotypes",
+        "isAbstract": false,
+        "members": [],
+        "extends_": [],
+        "implements_": [],
+        "generics": "",
+        "stereotype": "S1"
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:52]: should be: `[ "S1", "S2", "S3" ]`

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      },
+      {
+        "name": "TestStereotypeWithDoubleAngleBrackets",
+        "title": "TestStereotypeWithDoubleAngleBrackets",
+        "isAbstract": false,
+        "members": [],
+        "extends_": [],
+        "implements_": [],
+        "generics": "",
+        "stereotype": "<<S1"
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:72]: weirdly, this is correct like in the parsed diagram. But maybe we should not support this?

File: src/plantuml.pegjs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Enteee) [reply src/plantuml.pegjs:506]: did a few more experiments, see my next review.

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 10]
```diff
@@ -0,0 +1,687 @@
...
+            ]
+          ],
+          {
+            "name": "TestGenericsWithAngleBrackets",
+            "title": "TestGenericsWithAngleBrackets",
+            "isAbstract": false,
+            "members": [],
+            "extends_": [],
+            "implements_": [],
+            "generics": "<G1",
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:251]: should be `"<G1>"`

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 11]
```diff
@@ -0,0 +1,687 @@
...
+            "isAbstract": false,
+            "members": [],
+            "extends_": [],
+            "implements_": [],
+            "generics": "G1",
+            "stereotype": ""
+          },
+          [
+            [],
+            "'",
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:21]: this is a comment, and should be parsed as such. But I would ignore this for that pull request. Will fix with: https://github.com/Enteee/plantuml-parser/issues/46

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 12]
```diff
@@ -0,0 +1,687 @@
...
+            "stereotype": ""
+          },
+          {
+            "name": "TestGenericsWithDoubleAngleBrackets",
+            "title": "TestGenericsWithDoubleAngleBrackets",
+            "isAbstract": false,
+            "members": [],
+            "extends_": [],
+            "implements_": [],
+            "generics": "<<G1",
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:261]: should be: `"<<G1>>"`

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 13]
```diff
@@ -0,0 +1,687 @@
...
+          },
+          {
+            "name": "TestMultipleStereotypes",
+            "title": "TestMultipleStereotypes",
+            "isAbstract": false,
+            "members": [],
+            "extends_": [],
+            "implements_": [],
+            "generics": "",
+            "stereotype": "S1"
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:289]: should be : `"stereotype": ["S1", "S2", "S3"]`

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 14]
```diff
@@ -0,0 +1,687 @@
...
+          },
+          {
+            "name": "TestStereotypeWithDoubleAngleBrackets",
+            "title": "TestStereotypeWithDoubleAngleBrackets",
+            "isAbstract": false,
+            "members": [],
+            "extends_": [],
+            "implements_": [],
+            "generics": "",
+            "stereotype": "<<S1"
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:309]: weirdly this is correct.

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 15]
```diff
@@ -0,0 +1,687 @@
...
+          },
+          {
+            "name": "TestGenericsWithDoubleAngleBracketsAndStereotype",
+            "title": "TestGenericsWithDoubleAngleBracketsAndStereotype",
+            "isAbstract": false,
+            "members": [],
+            "extends_": [],
+            "implements_": [],
+            "generics": "<<G1",
+            "stereotype": ""
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:651]: should be:
```
            "generics": "<<G1>>",
            "stereotype": ["S1"]
```

File: test/fixtures/class-generics-and-stereotypes/parse-out.default [hunk 16]
```diff
@@ -0,0 +1,687 @@
...
+          },
+          {
+            "name": "TestGenericsAndMultipleStereotypes",
+            "title": "TestGenericsAndMultipleStereotypes",
+            "isAbstract": false,
+            "members": [],
+            "extends_": [],
+            "implements_": [],
+            "generics": "G1",
+            "stereotype": "S1"
```
Comment (Enteee) [inline test/fixtures/class-generics-and-stereotypes/parse-out.default:661]: should be: `["S1", "S2", "S3"]`

File: test/fixtures/class-generics-and-stereotypes/parse-out.default (hunk 11 above)
Comment (Enteee) [reply test/fixtures/class-generics-and-stereotypes/parse-out.default]: fixed

**Comment by paduc** [comment]
@Enteee indeed I didn't know that generics and stereotypes could include angle brackets like that. I guess we can find a way to handle that in the parser. Should handling these special cases really block this PR though ? I'm thinking special cases could be handled in later iterations and another PR. 

There's one other case I think would be interesting to parse (in yet another PR) : `MyClass <a extends b>`. The result after this PR would be `generics: "a extends b"` but maybe it would be nice to have `generics: [{ name: "a", extends: ["b"]}]`.

**Comment by Enteee** [comment]
> @Enteee indeed I didn't know that generics and stereotypes could include angle brackets like that. I guess we can find a way to handle that in the parser. Should handling these special cases really block this PR though ? I'm thinking special cases could be handled in later iterations and another PR.

Yes I agree, and I also do think its ok when we don't implement angle brackets handling for generics and stereotypes with this pull request. I will instead just open a bug report. 

> There's one other case I think would be interesting to parse (in yet another PR) : `MyClass <a extends b>`. The result after this PR would be `generics: "a extends b"` but maybe it would be nice to have `generics: [{ name: "a", extends: ["b"]}]`.

Honestly, I would not implement this. Especially not in this PR. If you do think there is value in doing this feel free to open an enhancement request. Where we then can discuss this.

Nevertheless, there is still one thing that I would like to see changed in this PR: It is possible to declare multiple stereotypes:

```
class TestMultipleStereotypes <<S1>> <<S2>> <<S3>>
```

Therefore the stereotype property should become a list of strings instead of just a string. And the grammar should be changed so that it allows multiple stereotypes. Do you think you can do this or should I have a look?

**Comment by Enteee** [comment]
opened the bug for angle brackets handling: https://github.com/Enteee/plantuml-parser/issues/59

File: test/fixtures/class-generics-and-stereotypes/parse-out.default (hunk 10 above)
Comment (Enteee) [reply test/fixtures/class-generics-and-stereotypes/parse-out.default:29]: handled in bug: https://github.com/Enteee/plantuml-parser/issues/59

File: test/fixtures/class-generics-and-stereotypes/parse-out.default (hunk 12 above)
Comment (Enteee) [reply test/fixtures/class-generics-and-stereotypes/parse-out.default:39]: handled in bug: https://github.com/Enteee/plantuml-parser/issues/59

File: test/fixtures/class-generics-and-stereotypes/parse-out.default (hunk 15 above)
Comment (Enteee) [reply test/fixtures/class-generics-and-stereotypes/parse-out.default:127]: handled in bug: https://github.com/Enteee/plantuml-parser/issues/59

**Comment by paduc** [comment]
> Therefore the stereotype property should become a list of strings instead of just a string. And the grammar should be changed so that it allows multiple stereotypes. Do you think you can do this or should I have a look?

I agree ! I’ll have a look ASAP.

**Comment by Enteee** [comment]
@paduc how is it going with this? do you need any help?

**Comment by paduc** [comment]
> @paduc how is it going with this? do you need any help?

Hello @Enteee, I'm on vacation right now. I'll get back to this in a few weeks.

**Comment by Enteee** [comment]
@paduc ok, no worries. Enjoy your holidays :airplane:

**Pull request merged**